core/arch/arm/kernel/mutex.c

```diff
@@ -81,6 +110,27 @@ static void __mutex_unlock(struct mutex *m, const char *fname, int lineno)
 	wq_wake_next(&m->wq, m, fname, lineno);
 }
 
+static void __mutex_unlock_recursive(struct recursive_mutex *m,
+				     const char *fname, int lineno)
+{
+	assert_have_no_spinlock();
+	assert(m->owner == thread_get_id_may_fail());
+
+	if (refcount_dec(&m->lock_depth)) {
+		/*
+		 * Do an atomic store to match the atomic load in
+		 * __mutex_lock_recursive()
+		 */
+		atomic_store_short(&m->owner, THREAD_ID_INVALID);
+		/* Proceed with unlock below */
+	} else {
+		/* The mutex is still write locked by this thread */
+		return;
+	}
+
+	__mutex_unlock(&m->m, fname, lineno);
```

Comment:
This function could be simplified by moving this up to the line below the `atomic_store_short()` above.